guides/keras_cv/simsiam-with-kerascv.py

```diff
@@ -0,0 +1,604 @@
+        kernel_initializer=tf.keras.initializers.LecunUniform(),
+        name="predictor_output",
+    )(x)
+    # Metric Logging layer. Monitors the std of the layer activations.
+    # Degnerate solutions colapse to 0 while valid solutions will move
+    # towards something like 0.0220. The actual number will depend on the layer size.
+    o = tfsim.layers.ActivationStdLoggingLayer(name="pred_std")(x)
+    predictor = tf.keras.Model(inputs, o, name="predictor")
+    return predictor
+
+
+predictor = get_predictor(input_dim=DIM, hidden_dim=512)
+predictor.summary()
+
+
+"""
+## Training
+
+First, we need to initialize our training model, loss, and optimizer.
+"""
+loss = tfsim.losses.SimSiamLoss(projection_type="cosine_distance", name="simsiam")
+
+contrastive_model = tfsim.models.ContrastiveModel(
+    backbone=backbone,
+    projector=projector,
+    predictor=predictor,  # NOTE: simiam requires predictor model.
+    algorithm="simsiam",
+    name="simsiam",
+)
+lr_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=INIT_LR,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+wd_decayed_fn = tf.keras.optimizers.schedules.CosineDecay(
+    initial_learning_rate=WEIGHT_DECAY,
+    decay_steps=PRE_TRAIN_EPOCHS * PRE_TRAIN_STEPS_PER_EPOCH,
+)
+optimizer = tfa.optimizers.SGDW(
+    learning_rate=lr_decayed_fn, weight_decay=wd_decayed_fn, momentum=0.9
+)
+
+"""
+Next we can compile the model the same way you compile any other Keras model.
+"""
+
+contrastive_model.compile(
+    optimizer=optimizer,
+    loss=loss,
+)
+
+"""
+We track the training using several callbacks.
+
+* **EvalCallback** creates an index at the end of each epoch and provides a proxy for the nearest neighbor matching classification using `binary_accuracy`.
+* **TensordBoard** and **ModelCheckpoint** are provided for tracking the training progress.
+"""
+
+DATA_PATH = Path("./")
+log_dir = DATA_PATH / "models" / "logs" / f"{loss.name}_{time.time()}"
+chkpt_dir = DATA_PATH / "models" / "checkpoints" / f"{loss.name}_{time.time()}"
+
+callbacks = [
+    tfsim.callbacks.EvalCallback(
+        tf.cast(x_query, tf.float32),
+        y_query,
+        tf.cast(x_index, tf.float32),
+        y_index,
+        metrics=["binary_accuracy"],
+        k=1,
+        tb_logdir=log_dir,
+    ),
+    tf.keras.callbacks.TensorBoard(
+        log_dir=log_dir,
+        histogram_freq=1,
+        update_freq=100,
+    ),
+    tf.keras.callbacks.ModelCheckpoint(
+        filepath=chkpt_dir,
+        monitor="val_loss",
+        mode="min",
+        save_best_only=True,
+        save_weights_only=True,
+    ),
+]
+
+"""
+All that is left to do is run fit()!
+"""
+
+print(train_ds)
+print(val_ds)
+history = contrastive_model.fit(
+    train_ds,
+    epochs=PRE_TRAIN_EPOCHS,
+    steps_per_epoch=PRE_TRAIN_STEPS_PER_EPOCH,
+    validation_data=val_ds,
+    validation_steps=VAL_STEPS_PER_EPOCH,
+    callbacks=callbacks,
+)
+
+
+"""
+## Plotting and Evaluation
```

Comment:
It would be nice to offer a utility function to plot learning curves.  I suppose I can create a nice looking one in `luketils`.